crates/containerd-shim-wasm/src/sandbox/youki_instance.rs

```diff
@@ -0,0 +1,188 @@
+//! Abstractions for running/managing a wasm/wasi instance that uses youki's libcontainer library.
+
+use anyhow::Context;
+use chrono::Utc;
+use libc::{SIGINT, SIGKILL};
+use libcontainer::{
+    container::{Container, ContainerStatus},
+    signal::Signal,
+};
+use log::error;
+use nix::{
+    errno::Errno,
+    sys::wait::{waitid, Id as WaitID, WaitPidFlag, WaitStatus},
+};
+
+use crate::sandbox::{
+    instance::ExitCode,
+    instance_utils::{get_instance_root, instance_exists},
+};
+
+use crate::sandbox::InstanceConfig;
+use std::{path::PathBuf, thread};
+
+use super::{error::Error, instance::Wait, Instance};
+
+/// YoukiInstance is a trait that gets implemented by a WASI runtime that
+/// uses youki's libcontainer library as the container runtime.
+/// It provides default implementations for some of the Instance trait methods.
+/// The implementor of this trait is expected to implement the
+/// * `new_youki()`
+/// * `get_exit_code()`
+/// * `get_id()`
+/// * `get_root_dir()`
+/// * `build_container()`
+/// methods.
+#[cfg(feature = "libcontainer")]
+pub trait YoukiInstance {
```

Comment:
**question:** should this be `LibcontainerInstance`? This is the first mention of `youki` AFAIK

Reply:
Done.